### PR TITLE
Improved hooks monitoring 

### PR DESCRIPTION
**Pull request by samc24**
...to work for specific edge cases with a simpler solution.
Re-reads hooks directories after any changes are detected by the watchers.
Added monitoring test for adding a different invalid hook to primary directory.
Some issues with prior code:
- `ReadDir` would stop when it encounters an invalid hook, rather than registering an error but continuing to read the valid hook.
- Wouldn’t account for Rename and Chmod events.
- After doing a `mv` of the hooks file instead of `rm`, it would still think the hooks file is in the directory, but it has been moved to another location.
- If a hook file was renamed, it would register the renamed file as a separate hook and not delete the original, so it would then execute the hook twice - once for the renamed file, and once for the original name which it did not delete.

Signed-off-by: samc24 <sam.chaturvedi24@gmail.com>

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@samc24 thanks for the PR.  I think you shortened the PR title in GitHub, but did not do `git commit --amend -s`, shorten it there too, then repush.  You'll need to do that to get the git validation happy.

File: pkg/hooks/monitor.go [hunk 1]
```diff
@@ -2,9 +2,8 @@ package hooks
 
 import (
 	"context"
-	"os"
-	"path/filepath"
 
+	current "github.com/containers/libpod/pkg/hooks/1.0.0"
```
Comment (TomSweeneyRedHat) [inline pkg/hooks/monitor.go]: suggest `libpodhook` over `current`, unless @mheon has a better suggestion.

Comment (mheon) [reply pkg/hooks/monitor.go]: I'd call it `hooks-cur` or `hooks-current`

Comment (TomSweeneyRedHat) [reply pkg/hooks/monitor.go]: Never mind, I see current is in the package.

**Comment by mheon** [comment]
Validate is failing on the first line of your commit being over 90 characters in length

**Comment by mheon** [comment]
I do like the simplification in the code a lot, though.

**Comment by samc24** [comment]
oops, messed something up while trying to fix and squash commits. ignore above, will try to fix that!

**Comment by rhatdan** [comment]
@saschagrunert I thought we had determined ReadDir was not working correctly and would fail differently then in the past.  Basically we need ReadDir to complain about bad hooks, but to continue loading the other hooks, before returning the error.

File: pkg/hooks/read.go [hunk 2]
```diff
@@ -80,12 +80,13 @@ func ReadDir(path string, extensionStages []string, hooks map[string]*current.Ho
 					continue
 				}
 			}
-			return err
+			res = err
```
Comment (vrothberg) [inline pkg/hooks/read.go]: This will only report the last error and we'll lose previous ones. Can we instead do as follows?
```golang
if res == nil {
   res = err
} else {
   res = errors.Wrapf(res, "%v", err)
}
```

Comment (vrothberg) [reply pkg/hooks/read.go]: `errors.Wrapf(nil, ...)` always returns nil, so we need to make it conditional.

File: pkg/hooks/monitor.go [hunk 3]
```diff
@@ -49,47 +48,11 @@ func (m *Manager) Monitor(ctx context.Context, sync chan<- error) {
...
-				if (found || event.Op&fsnotify.Remove == fsnotify.Remove) && err != nil {
-					ok := m.remove(filename)
-					if ok {
-						logrus.Debugf("removed hook %s (triggered by activity on %s)", filename, event.Name)
-					}
+			m.hooks = make(map[string]*current.Hook)
+			for _, dir := range m.directories {
+				err = ReadDir(dir, m.extensionStages, m.hooks)
+				if err != nil {
+					logrus.Errorf("failed %s: %v", event.Name, err)
```
Comment (vrothberg) [inline pkg/hooks/monitor.go]: Can we make the log a bit more verbose? E.g., `failed loading hooks for %s: %v`.

**Comment by vrothberg** [comment]
@rhatdan @TomSweeneyRedHat @giuseppe @saschagrunert PTAL

File: pkg/hooks/monitor.go (hunk 1 above)
Comment (rhatdan) [inline pkg/hooks/monitor.go]: This seems like a weird name, why not have it.

hooks "github.com/containers/libpod/pkg/hooks/1.0.0"

Comment (samc24) [reply pkg/hooks/monitor.go]: @rhatdan, agreed, but I used it to stay consistent with other usages in other files, where it is also named 'current'. Think I should change it here and everywhere else?

Comment (vrothberg) [reply pkg/hooks/monitor.go]: I suggest to stick with `current` in this PR. We can clean this up later and if desired open an issue to not forget about it.

**Comment by umohnani8** [comment]
LGTM
Tests are green! @mheon @vrothberg @rhatdan PTAL
